modules/src/ics07_tendermint/client_def.rs

```diff
@@ -25,17 +36,117 @@ impl ClientDef for TendermintClient {
+                    consensus_state,
+                ));
+            } else {
+                return Ok((client_state, consensus_state));
+            }
+        };
+
+        let latest_consensus_state =
+            match ctx.consensus_state(&client_id, client_state.latest_height) {
+                //could the header height be zero ?
+                Some(cs) => downcast!(
+                    cs => AnyConsensusState::Tendermint
+                )
+                .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?,
+                None => {
+                    return Err(Ics02Error::consensus_state_not_found(
+                        client_id.clone(),
+                        client_state.latest_height,
+                    ));
+                }
+            };
+
+        monotonicity_checks(latest_consensus_state, header.clone(), client_state.clone())
+            .map_err(Ics02Error::tendermint_handler_error)?;
 
-        // TODO: Additional verifications should be implemented here.
+        // check that the versions of the client state and the header match
+        if client_state.latest_height.revision_number != header.height().revision_number {
+            return Err(Error::mismatched_revisions(
+                client_state.latest_height.revision_number,
+                header.height().revision_number,
+            )
+            .into());
+        };
+
+        let trusted_consensus_state = match ctx.consensus_state(&client_id, header.trusted_height) {
+            Some(ts) => downcast!(
+                ts => AnyConsensusState::Tendermint
+            )
+            .ok_or_else(|| Ics02Error::client_args_type_mismatch(ClientType::Tendermint))?,
+            None => {
+                return Err(Ics02Error::consensus_state_not_found(
+                    client_id,
+                    header.trusted_height,
+                ))
+            }
+        };
+
+        if header.height() == header.trusted_height.increment() {
+            //adjacent
+
+            // check that the header's trusted validator set is
+            // the next_validator_set of the trusted consensus state
+            if Set::hash(&header.validator_set) != trusted_consensus_state.next_validators_hash {
+                return Err(Error::invalid_validator_set(
+                    trusted_consensus_state.next_validators_hash,
+                    Set::hash(&header.validator_set),
+                )
+                .into());
+            }
+
+            // check that the validators that sign the commit of the untrusted header
+            // have 2/3 of the voting power of the current validator set.
+            if let Err(e) = voting_power_in(
+                &header.signed_header,
+                &header.validator_set,
+                TrustThresholdFraction::TWO_THIRDS,
+            ) {
+                return Err(Error::insufficient_voting_power(e.to_string()).into());
+            }
```

Comment:
same here, the check should happen for both adjacent and non-adjacent headers. In fact you have this below in the `else` branch.